sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala

```diff
@@ -86,7 +86,7 @@ trait FileScan extends Scan
 
   override def equals(obj: Any): Boolean = obj match {
     case f: FileScan =>
-      fileIndex == f.fileIndex && readSchema == f.readSchema
+      fileIndex == f.fileIndex && readSchema == f.readSchema &&
         ExpressionSet(partitionFilters) == ExpressionSet(f.partitionFilters) &&
         ExpressionSet(dataFilters) == ExpressionSet(f.dataFilters)
```

Comment:
shall we canonicalize the filters here to fix the exchange reuse issue?

Reply:
But don't we need the `output` of the ScanExec (`BatchScanExec`) node to do that?

Reply:
You are right. Then it's hard to canonicalize `Scan` implementations that are outside of Spark...

Reply:
Thinking about it a bit more, do we really need `output`? We can canonicalize expr IDs to 0 and only look at the column name.

Reply:
Yes, I was thinking about the same. I will try to do this today.

Reply:
I was thinking about a change like in https://github.com/apache/spark/pull/31848/commits/f65ebe3f397053e103d482c50d7f250f798bb825.
That way we don't need to duplicate some parts of the canonicalization logic defined in `QueryPlan.normalizeExpressions` and in `Expression.canonicalized` (`Canonicalize.expressionReorder`, `Canonicalize.ignoreTimeZone`)...

If this looks ok to you then I can add UTs for `.equals()` into a new `FileScanSuite` as @dongjoon-hyun suggested and move the e2e UT from https://github.com/apache/spark/pull/31820 to here as we don't need the other change in `BatchScanExec`.

Reply:
SGTM